sdk/digitaltwins/digital-twins-parser/src/parser/idValidator.ts

```diff
@@ -0,0 +1,162 @@
+      }
+      const length = IdValidator.idDefinitionMaxLengths.get(dtdlVersion);
+      if (length && idToken.length > length) {
+        parsingErrors.push(
+          createParsingError("dtmi:dtdl:parsingError:idTooLong", {
+            cause: `Identifier ${idToken} is too long -- length limit is ${length}.`,
+            action: `Select a shorter value for the identifier or trim current value to fewer than  ${length} characters.`,
+            value: idToken
+          })
+        );
+        throw new ParsingException(parsingErrors);
+      }
+      const regex = IdValidator.idDefinitionRegexPatterns.get(dtdlVersion);
+      if (regex && !regex.test(idToken)) {
+        parsingErrors.push(
+          createParsingError("dtmi:dtdl:parsingError:invalidId", {
+            cause: `Identifier ${idToken} is invalid.`,
+            action:
+              "Replace the identifier with a string that conforms to the DTMI syntax-- see https://github.com/Azure/digital-twin-model-identifier.",
+            value: idToken
+          })
+        );
+        throw new ParsingException(parsingErrors);
+      }
+      idString = idToken;
+    } else {
+      if (idRequired) {
+        parsingErrors.push(
+          createParsingError("dtmi:dtdl:parsingError:missingRequiredId", {
+            cause: `{primaryId:p} property ${propName} requires an identifer but none provided.`,
+            action:
+              'Add an "@id" property whose value is string that conforms to the DTMI syntax-- see https://github.com/Azure/digital-twin-model-identifier.',
+            primaryId: parentId,
+            property: propName
+          })
+        );
+        throw new ParsingException(parsingErrors);
+      }
+      let secondSeg = "";
+      if (dtmiSeg) {
+        const segToken = obj[dtmiSeg];
+        if (!segToken || typeof segToken !== "string") {
+          parsingErrors.push(
+            createParsingError("dtmi:dtdl:parsingError:missingRequiredProperty", {
+              cause: `{primaryId:p} property ${propName} requires property ${dtmiSeg} to be specified but it is not.`,
+              action: `Add a ${dtmiSeg} property with a string value that is unique across all values of ${propName}.`,
+              primaryId: parentId,
+              property: propName,
+              value: dtmiSeg
+            })
+          );
+          throw new ParsingException(parsingErrors);
+        }
+        secondSeg = ":__" + segToken;
+      }
+      if (parentId !== undefined) {
+        const delimeterPos = parentId.lastIndexOf(";");
+        if (delimeterPos < 0) {
+          idString = parentId + ":_" + propName + secondSeg;
+        } else {
+          idString =
+            parentId.substring(0, delimeterPos) +
+            ":_" +
+            propName +
+            secondSeg +
+            parentId.substring(delimeterPos);
+        }
+      }
+    }
+    return idString;
+  }
+  // codegen-outline-end
+}
+
+IdValidator.initialize();
```

Comment:
another strange pattern, why not just have the code run in module scope to do the setup and export the methods for parsing?